server/src/packages/rwcds-form/rwcds-form.service.ts

```diff
@@ -0,0 +1,40 @@
+import { Injectable, HttpStatus, HttpException } from '@nestjs/common';
+import { CrmService } from '../../crm/crm.service';
+import { pick } from 'underscore';
+import { RWCDS_FORM_ATTRS } from './rwcds-form.controller';
+
+@Injectable()
+export class RwcdsFormService {
+  constructor(private readonly crmService: CrmService) {}
+
+  async _getPackageProjectId(packageId):Promise<String> {
+    const { records: [firstPackage] } = await this.crmService.get('dcp_packages', `
+      $select=_dcp_project_value
+      &$filter=dcp_packageid eq ${packageId}
+    `);
+
+    if (!firstPackage) {
+      return undefined;
+    }
+
+    const { _dcp_project_value } = firstPackage;
+
+    return _dcp_project_value;
+  }
+
+  async update(id, body) {
+    const allowedAttrs = pick(body, RWCDS_FORM_ATTRS);
+    const { package: packageId } = body;
+    const projectId = await this._getPackageProjectId(packageId);
+
+    if(!projectId) {
+      return new HttpException('Package not found for RWCDS Form', HttpStatus.NOT_FOUND);
+    }
+
+    await this.crmService.update('dcp_rwcdsforms', id, allowedAttrs);
+
+    await this.crmService.update('dcp_projects', projectId, {
+      'dcp_anticipatedyearbuilt': allowedAttrs.dcp_buildyear,
+    });
```

Comment:
Why does this need to happen every time we update something in the rwcds form? Can this not happen when we update the project model? Can we make an update method on the project service? Or can a separate call be made to the projects controller? What about CRM — do we know that CRM doesn't not synchronize these things?

I'm not sure it makes sense to include an _update to a property on a project_ in the rwcds. If i didn't know anything about this code, I would have no idea why every single update to the rwcds form should require an update to the related project's dcp_anticipatedyearbuilt.

Also, it's a patch request. So what if the body doesn't include `dcp_anticipatedyearbuilt`? Does it get set to undefined? Or worse, does it error out because it doesn't allow null for some reason?

I think making the rwcds service was definitely the right move, but the one-off look of `dcp_anticipatedyearbuilt` is confusing to me. Let's talk approach because this won't be the first time we'll run into this.

Otherwise, great work, _very concise_ — I know this is an awkward requirement, and I think this is a great start.

Reply:
Per our standup discussion, I'll look into:
  - Whether RwcdsForm.buildyear is NULL if user does not modify the buildyear
  - Using ProjectService instead of CrmService to patch projects

Reply:
- I've confirmed that RwcdsForm.buildyear _isn't_ NULL if user doesn't modify the buildyear
- The ProjectService doesn't yet have any update() methods (we haven't been writing to Project yet). @allthesignals Do you think I should just add one?

Reply:
@godfreyyeung 

> * I've confirmed that RwcdsForm.buildyear _isn't_ NULL if user doesn't modify the buildyear

It doesn't matter. According to the JSON:API spec, we can't guarantee that all the attributes will be provided in the request body. Did you read the link I sent you?

Just because the FE application makes a request and it "works" doesn't make it a good design. The point is to _follow the spec as written_. Your code here violates the spec by overwriting the value in CRM if the body doesn't happen to include the attribute. Please fix it. If you are still confused about this please give me a call. Thanks.

Reply:
Per standup this morning, the crux of the matter is that, even though Applicant Portal FE always send RwcdsForm.buildyear, another consumer of the API may not. If buildyear is not a property in the request body, the backend will interpret it as NULL when it shouldn't be. I'll write a guard for this.